File: Algorithms/Python/0106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal_Solution_1.py

# Space: O(n)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, inorder, postorder):
        if len(inorder) == 0:
            return None

        root_value = postorder.pop()
        root = TreeNode(root_value)
        root_index_in_inorder = inorder.index(root_value)

        # CAUTION: we need to recursively build right node first.
        #          since we pop item from end of postorder list, so we will get right node first, then left node.
        root.right = self.buildTree(inorder[root_index_in_inorder + 1:], postorder)
        root.left = self.buildTree(inorder[:root_index_in_inorder], postorder)

        return root

File: Algorithms/0106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/Java/ConstructBinaryTreefromInorderandPostorderTraversalSolution1.java

// Space: O(n)
// Time: O(n)


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder.length == 0) return null;
        if (inorder.length == 1) return new TreeNode(inorder[0]);
        postorderIndex = postorder.length - 1;
        return constructTree(inorder,postorder);
    }

    private int postorderIndex;

    private TreeNode constructTree(int[] inorder, int[] postorder) {
        if (postorderIndex < 0 || inorder.length == 0 || postorder.length == 0) return null;
        int rootVal = postorder[postorderIndex--];
        TreeNode root = new TreeNode(rootVal);
        int inorderIndex = findIndex(inorder, rootVal);
        if (inorderIndex != -1){
            root.right = constructTree(Arrays.copyOfRange(inorder, inorderIndex + 1, inorder.length), postorder);
            root.left = constructTree(Arrays.copyOfRange(inorder, 0, inorderIndex), postorder);
        }
        return root;
    }

    private int findIndex(int[] array, int target) {
        for (int i = 0; i < array.length; i++) {
            if (array[i] == target) return i;
        }
        return -1;
    }
}




